Submitted By: Robert Connolly <robert at linuxfromscratch dot org> (ashes)
Date: 2004-04-18
Initial Package Version: 2.6.5
Origin: http://www.northernsecurity.net/adamantix/diffs/2.6.0-propolice.patch
Description: Linux kernel patch for Smashing Stack Protector. This patch allows
the use of -fstack-protector durring the kernel build.
http://www.linuxfromscratch.org/hints/downloads/files/ssp.txt
http://www.linuxfromscratch.org/hlfs/

diff -Naur linux-2.6.5.orig/Makefile linux-2.6.5.ssp/Makefile
--- linux-2.6.5.orig/Makefile	2004-04-04 03:37:36.000000000 +0000
+++ linux-2.6.5.ssp/Makefile	2004-04-19 01:52:26.000000000 +0000
@@ -287,7 +287,7 @@
 		   $(if $(KBUILD_SRC),-Iinclude2 -I$(srctree)/include)
 
 CFLAGS 		:= -Wall -Wstrict-prototypes -Wno-trigraphs \
-	  	   -fno-strict-aliasing -fno-common
+	  	   -fno-strict-aliasing -fno-common -fstack-protector
 AFLAGS		:= -D__ASSEMBLY__
 
 export	VERSION PATCHLEVEL SUBLEVEL EXTRAVERSION KERNELRELEASE ARCH \
diff -Naur linux-2.6.5.orig/arch/i386/kernel/i386_ksyms.c linux-2.6.5.ssp/arch/i386/kernel/i386_ksyms.c
--- linux-2.6.5.orig/arch/i386/kernel/i386_ksyms.c	2004-04-04 03:38:23.000000000 +0000
+++ linux-2.6.5.ssp/arch/i386/kernel/i386_ksyms.c	2004-04-19 01:52:26.000000000 +0000
@@ -92,6 +92,11 @@
 EXPORT_SYMBOL_NOVERS(__down_failed_interruptible);
 EXPORT_SYMBOL_NOVERS(__down_failed_trylock);
 EXPORT_SYMBOL_NOVERS(__up_wakeup);
+
+/* SSP */
+EXPORT_SYMBOL_NOVERS(__guard);
+EXPORT_SYMBOL_NOVERS(__stack_smash_handler);
+
 /* Networking helper routines. */
 EXPORT_SYMBOL(csum_partial_copy_generic);
 /* Delay loops */
diff -Naur linux-2.6.5.orig/arch/um/kernel/ksyms.c linux-2.6.5.ssp/arch/um/kernel/ksyms.c
--- linux-2.6.5.orig/arch/um/kernel/ksyms.c	2004-04-04 03:36:16.000000000 +0000
+++ linux-2.6.5.ssp/arch/um/kernel/ksyms.c	2004-04-19 01:52:26.000000000 +0000
@@ -90,3 +90,5 @@
 EXPORT_SYMBOL(kmap_atomic_to_page);
 #endif
 
+EXPORT_SYMBOL_NOVERS(__guard);
+EXPORT_SYMBOL_NOVERS(__stack_smash_handler);
diff -Naur linux-2.6.5.orig/include/linux/kernel.h linux-2.6.5.ssp/include/linux/kernel.h
--- linux-2.6.5.orig/include/linux/kernel.h	2004-04-04 03:36:12.000000000 +0000
+++ linux-2.6.5.ssp/include/linux/kernel.h	2004-04-19 01:52:26.000000000 +0000
@@ -118,6 +118,8 @@
 #define TAINT_FORCED_RMMOD		(1<<3)
 
 extern void dump_stack(void);
+extern int __guard;
+extern void __stack_smash_handler(int, char []);
 
 #ifdef DEBUG
 #define pr_debug(fmt,arg...) \
diff -Naur linux-2.6.5.orig/lib/Makefile linux-2.6.5.ssp/lib/Makefile
--- linux-2.6.5.orig/lib/Makefile	2004-04-04 03:36:16.000000000 +0000
+++ linux-2.6.5.ssp/lib/Makefile	2004-04-19 01:52:26.000000000 +0000
@@ -6,7 +6,7 @@
 lib-y := errno.o ctype.o string.o vsprintf.o cmdline.o \
 	 bust_spinlocks.o rbtree.o radix-tree.o dump_stack.o \
 	 kobject.o idr.o div64.o parser.o int_sqrt.o \
-	 bitmap.o extable.o
+	 bitmap.o extable.o propolice.o
 
 # hack for now till some static code uses krefs, then it can move up above...
 obj-y += kref.o
diff -Naur linux-2.6.5.orig/lib/propolice.c linux-2.6.5.ssp/lib/propolice.c
--- linux-2.6.5.orig/lib/propolice.c	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.6.5.ssp/lib/propolice.c	2004-04-19 01:52:26.000000000 +0000
@@ -0,0 +1,13 @@
+ /*
+  *  linux/lib/errno.c
+  *
+  *
+  */
+ 
+ int __guard = '\0\0\n\777';
+ 
+ void __stack_smash_handler (int damaged, char func[])
+ 	 {
+	   static char *message = "propolice detects %x at function %s.\n" ;
+	   panic (message, damaged, func);
+}
